tesseract_kinematics/core/include/tesseract_kinematics/core/utils.h

```diff
@@ -502,5 +502,45 @@ inline void harmonizeTowardZero(Eigen::Ref<VectorX<FloatType>> qs,
   }
 }
 
+/**
+ * @brief Check to see if the solution is within joint limits, and if not applies a redundant solution
+ * that is within joint limits if available.
+ * 
+ * @tparam FloatType 
+ * @param qs A pointer to a float array
+ * @param position_limits The position limits of the robot
+ */
+template <typename FloatType>
+inline void harmonizeRedundantSolutions(Eigen::Ref<VectorX<FloatType>> qs,
+                                const std::vector<Eigen::Index>& redundancy_capable_joints,
+                                const Eigen::Ref<const Eigen::Matrix<FloatType, Eigen::Dynamic, 2>>& position_limits)
+{
+  const static auto pi = FloatType(M_PI);
+  const static auto two_pi = FloatType(2.0 * M_PI);
+
+  for (const auto& i : redundancy_capable_joints)
+  {
+    // TODO: Should this harmonize here? It may have unpredictable results on numerical solvers
+    // FloatType diff = std::fmod(qs[i] + pi, two_pi);
+    // qs[i] = (diff < 0) ? (diff + pi) : (diff - pi);    
+    if (qs[i] < position_limits(i,0) && position_limits(i,0) > -pi)
+    {
+      double qs_i_2pi = two_pi + qs[i];
+      if (qs_i_2pi < position_limits(i,1))
+      {
+        qs[i] = qs_i_2pi;
+      }
+    }
+    else if (qs[i] > position_limits(i,1) && position_limits(i,1) < pi)
+    {
+      double qs_i_2pi = qs[i] - two_pi;
+      if (qs_i_2pi > position_limits(i,0))
+      {
+        qs[i] = qs_i_2pi;
+      }
+    }        
```

Comment:
> Limits [-12pi , -8pi], where value is zero would it give a solution between the limits?

No, this assumes that 0 is within the joint limits. I have never seen a robot with joint limits that don't cross 0, the issue is if the limits don't cover [-pi,pi]

Reply:
We will need to make sure these use cases are handled, because I have used several custom robot which which do have limits that don't cross the 0.

Reply:
Would it be better to harmonize relative to the mean of the limits?

Reply:
> We will need to make sure these use cases are handled because I have used several custom robot which do have limits that don't cross the 0.

Can you share an example URDF of a robot like this?

Reply:
I think harmonizing to the mean might be a better approach. Below is example code which seems to work.

``` c++
double harmonize(double v, double mean)
{
  const static auto pi = double(M_PI);
  const static auto two_pi = double(2.0 * M_PI);

  const double mean_diff = std::fmod(mean + pi, two_pi);
  const double diff = std::fmod(v + pi, two_pi);

  const double mv = (mean_diff < 0) ? (mean_diff + pi) : (mean_diff - pi);
  const double vv = (diff < 0) ? (diff + pi) : (diff - pi);

  return (mean + (vv - mv));
}
```

Reply:
I have typically solved for the solution closest to zero on other situations. The solution here is intended to have the smallest performance impact. Maybe have another nested block that checks if there are even stranger limits to avoid doing floating point math unless necessary?

Reply:
![image](https://github.com/tesseract-robotics/tesseract/assets/9803128/0d71cfed-43dd-402f-918d-9830320314d2)